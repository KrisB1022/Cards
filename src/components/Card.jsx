import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { css } from "emotion";

import {
  Button,
  Card as ReactStrapCard,
  CardHeader,
  CardBody,
  CardImg,
  CardText
} from "reactstrap";

const imageStyles = css`
  max-width: 200px;
`;

class Card extends PureComponent {
  cardTextMarkup = ({ label, text }) => {
    return (
      <CardText>
        <span className="font-weight-bolder">{`${label}:`}</span> {text}
      </CardText>
    );
  };

  addToUserCards = () => {
    const { card, updateUserCards } = this.props;
    updateUserCards({ card });
  };

  removeFromUserCards = () => {
    const { card, updateUserCards } = this.props;
    updateUserCards({ card, remove: true });
  };

  render() {
    const { card, isUserCard } = this.props;
    const { imageUrl, name, set, text } = card;

    return (
      <div className="col-md-3 flex-grow-1 mb-4">
        <ReactStrapCard className="pl-0 pr-0 h-100">
          <CardHeader>{name}</CardHeader>

          <CardBody className="d-flex flex-column">
            <figure className="figure mx-auto d-block mb-auto border-bottom pb-2">
              <CardImg
                top
                className={`mx-auto d-block figure-img img-fluid ${imageStyles} flex-grow-1`}
                src={imageUrl}
                alt={`${name} card image`}
              />
              <figcaption className="figure-caption text-right">
                {name}
              </figcaption>
            </figure>

            <div className="pt-2 pb-2">
              {text &&
                this.cardTextMarkup({
                  label: "Card info",
                  text
                })}

              {set &&
                this.cardTextMarkup({
                  label: "Set Name",
                  text: set.name
                })}
            </div>

            {isUserCard ? (
              <Button color="secondary" onClick={this.removeFromUserCards}>
                Remove from my cards
              </Button>
            ) : (
              <Button color="primary" onClick={this.addToUserCards}>
                Add to my cards
              </Button>
            )}
          </CardBody>
        </ReactStrapCard>
      </div>
    );
  }
}

Card.propTypes = {
  card: PropTypes.object.isRequired,
  isUserCard: PropTypes.bool
};

Card.defaultProps = {
  imageUrl: "https://via.placeholder.com/223x310.png"
};

export default Card;
