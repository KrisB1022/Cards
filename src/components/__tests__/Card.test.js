import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import { Card as ReactStrapCard, CardHeader, CardBody, CardImg, CardText } from "reactstrap";
import Card from "../Card";

let wrapper;

beforeEach(() => {
	wrapper = shallow(<Card imageUrl="image.com" name="name" />);
});

it("renders without crashing", () => {
	shallow(<Card imageUrl="image.com" name="name" />);
});

it("has 1 <ReactStrapCard/>", () => {
	expect(wrapper.find(ReactStrapCard).exists()).toBeTruthy();
	expect(wrapper.find(ReactStrapCard).length).toBe(1);
});

it("has 1 <CardBody/>", () => {
	expect(wrapper.find(CardBody).exists()).toBeTruthy();
	expect(wrapper.find(CardBody).length).toBe(1);
});

it("has 1 <CardImg/>", () => {
	expect(wrapper.find(CardImg).exists()).toBeTruthy();
	expect(wrapper.find(CardImg).length).toBe(1);
});

it("has 3 <CardText/>", () => {
	wrapper.setProps({
		artist: "artist",
		originalType: "originalType",
		setName: "setName"
	});

	expect(wrapper.find(CardText).exists()).toBeTruthy();
	expect(wrapper.find(CardText).length).toBe(3);
});

it("matches snapshot", () => {
	const tree = renderer.create(
		<Card
			imageUrl="url.com"
			name="name"
			artist="artist"
			originalType="originalType"
			setName="setName"
		/>
	);

	expect(tree.toJSON()).toMatchSnapshot();
});
