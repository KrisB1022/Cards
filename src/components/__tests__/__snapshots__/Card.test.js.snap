// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
.emotion-0 {
  max-width: 200px;
}

<div
  className="col-md-3 flex-grow-1 mb-4"
>
  <div
    className="pl-0 pr-0 h-100 card"
  >
    <div
      className="card-header"
    >
      a card!
    </div>
    <div
      className="d-flex flex-column card-body"
    >
      <figure
        className="figure mx-auto d-block mb-auto border-bottom pb-2"
      >
        <img
          alt="a card! card image"
          className="mx-auto d-block figure-img img-fluid emotion-0 flex-grow-1 card-img-top"
          src="someimage.com"
        />
        <figcaption
          className="figure-caption text-right"
        >
          a card!
        </figcaption>
      </figure>
      <div
        className="pt-2 pb-2"
      >
        <p
          className="card-text"
        >
          <span
            className="font-weight-bolder"
          >
            Card info:
          </span>
           
          Here is what this card does
        </p>
        <p
          className="card-text"
        >
          <span
            className="font-weight-bolder"
          >
            Set Name:
          </span>
           
          I belong to this set
        </p>
      </div>
      <button
        aria-label={null}
        className="btn btn-primary"
        onClick={[Function]}
        type="button"
      >
        Add to my cards
      </button>
    </div>
  </div>
</div>
`;
